File: day8/cpptest/day8.cpp
#include <cctype>
#include <cstdio>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

struct Destinations {
    Destinations() {
        this->left = "";
        this->right = "";
    }
    Destinations(string left, string right) {
        this->left = left;
        this->right = right;
    }
    string left;
    string right;
};

vector<string> split(const string &s, char delimeter) {
    vector<string> result;
    stringstream ss(s);
    string item;

    while (getline(ss, item, delimeter)) {
        if (item.length() <= 0) {
            continue;
        } 
        result.push_back(item);
    }

    return result;
} 

string trimSpace(const string &originalStr) {
    string trimmed = "";
    size_t firstNonSpace = 0;
    size_t lastNonSpace = originalStr.size() - 1;

    while (true) {
        if (originalStr[firstNonSpace] != ' ') {
            break;
        }
        firstNonSpace++;
    }
    while (true) {
        if (originalStr[lastNonSpace] != ' ') {
            break;
        }
        lastNonSpace--;
    }

    for (size_t i = firstNonSpace; i <= lastNonSpace; i++) {
        trimmed += originalStr[i]; 
    }

    return trimmed;
}

pair<string, string> getDestinations(const string &dests) {
    string left = "";
    string right = "";

    vector<string> destsSplit = split(dests, ',');
    for (size_t i = 0; i < destsSplit[0].size(); i++) {
        if (isalnum(destsSplit[0][i])) {
            left += destsSplit[0][i];
        }
    }
    for (size_t i = 0; i < destsSplit[1].size(); i++) {
        if (isalnum(destsSplit[1][i])) {
            right += destsSplit[1][i];
        }
    }

    return make_pair(left, right);
}

unordered_map<string, Destinations> parseLines(const vector<string> &translations) {
    unordered_map<string, Destinations> destMap;
    for (const string& bufline : translations) {
        vector<string> buflineSplit = split(bufline, '=');
        string key = trimSpace(buflineSplit[0]);
        string dests = trimSpace(buflineSplit[1]);
        pair<string, string> destPair = getDestinations(dests);
        Destinations destStruct = Destinations(destPair.first, destPair.second);
        destMap[key] = destStruct;
    }

    return destMap;
}

vector<string> getStartPositions(const unordered_map<string, Destinations> &destMap) {
    vector<string> startPositions;
    for (auto it = destMap.begin(); it != destMap.end(); it++) {
        string position = it->first;
        if (position[position.size() - 1] == 'A') {
            startPositions.push_back(position);
        }
    }

    return startPositions;
}

bool allOnZ(vector<string> &positions) {
    for (const string& pos : positions) {
        if (pos[pos.size() - 1] != 'Z') {
            return false;
        }
    }
    return true;
}

template <typename T>
void show(vector<T> v) {
    for (const T& t : v) {
        cout << t << " ";
    }
    printf("\n");
}

int getNumSteps(
    const string &steps,
    const unordered_map<string, Destinations> &destMap,
    const vector<string> &startPositions
) {
    int numSteps = 0;
    vector<string> positions(startPositions);

    while (!allOnZ(positions)) {
        int index = numSteps % steps.size();
        numSteps++;
        char direction = steps[index];
        switch (direction) {
            case 'L': {
                for (size_t i = 0; i < positions.size(); i++) {
                    string key = positions[i];
                    positions[i] = destMap.find(key)->second.left;
                }
                break;
            }
            case 'R': {
                for (size_t i = 0; i < positions.size(); i++) {
                    string key = positions[i];
                    positions[i] = destMap.find(key)->second.right;
                }
                break;
            }
            default: continue;
        }
        printf("%d ", numSteps);
        show(positions);
    }
    
    return numSteps;
}

int main() {
    ifstream inputfile("d8");
    string line;
    vector<string> buffer;
    while (getline(inputfile, line)) {
        buffer.push_back(line);
    }

    string steps = buffer[0];
    vector<string>::iterator sliceStart = buffer.begin() + 2;
    vector<string>::iterator sliceEnd = buffer.end();
    vector<string> translations(sliceStart, sliceEnd);

    unordered_map<string, Destinations> destMap = parseLines(translations);
    vector<string> startPositions = getStartPositions(destMap);

    int numSteps = getNumSteps(steps, destMap, startPositions);
    printf("%d\n", numSteps);
}

File: day8/day8.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type Destinations struct {
    left string
    right string
}

func parseLines(buffer []string) map[string]Destinations {
    destMap := make(map[string]Destinations)
    for _, bufline := range(buffer) {
        buflineSplit := strings.Split(bufline, " = ")
        key := buflineSplit[0]
        dests := string([]rune(buflineSplit[1])[1:len(buflineSplit[1])-1])
        destsSplit := strings.Split(dests, ", ")
        destStruct := Destinations{
            left: destsSplit[0],
            right: destsSplit[1],
        }

        destMap[key] = destStruct
    }

    return destMap
}

func parseLines2(buffer []string) (map[string]Destinations, []string) {
    destMap := make(map[string]Destinations)
    var startPositions []string
    for _, bufline := range(buffer) {
        buflineSplit := strings.Split(bufline, " = ")
        key := buflineSplit[0]
        dests := string([]rune(buflineSplit[1])[1:len(buflineSplit[1])-1])
        destsSplit := strings.Split(dests, ", ")
        destStruct := Destinations{
            left: destsSplit[0],
            right: destsSplit[1],
        }

        destMap[key] = destStruct

        if []rune(key)[len(key) - 1] == 'A' {
            startPositions = append(startPositions, key)
        }
    }

    return destMap, startPositions
}

func getNumSteps(steps string, destMap map[string]Destinations) int {
    curr := "AAA"
    numSteps := 0
    stepsArray := []rune(steps)
    for true {
        if curr == "ZZZ" {
            break
        }
        index := numSteps % len(stepsArray) 
        direction := stepsArray[index]
        if (direction == 'L') {
            curr = destMap[curr].left
        } else {
            curr = destMap[curr].right
        }
        numSteps += 1
    }

    return numSteps
}

func allOnZ(positions []string) bool {
    for _, pos := range(positions) {
        posArray := []rune(pos)
        if posArray[len(pos) - 1] != 'Z' {
            return false
        } 
    }
    return true
}

func getNumSteps2(steps string, destMap map[string]Destinations, startPositions []string) int {
    numSteps := 0
    stepsArray := []rune(steps)
    positions := startPositions
    for true {
        fmt.Println(positions, numSteps)
        if allOnZ(positions) {
            break
        }
        index := numSteps % len(stepsArray)
        direction := stepsArray[index]
        if direction == 'L' {
            for i, pos := range(positions) {
                positions[i] = destMap[pos].left
            }
        } else {
            for i, pos := range(positions) {
                positions[i] = destMap[pos].right
            }
        }
        numSteps += 1
    }

    return numSteps
}

func getNumSteps3(steps string, destMap map[string]Destinations, startPosition string) int {
    numSteps := 0
    stepsArray := []rune(steps)
    curr := startPosition;
    for true {
        if []rune(curr)[len(curr) - 1] == 'Z' {
            break
        }
        index := numSteps % len(stepsArray) 
        direction := stepsArray[index]
        if (direction == 'L') {
            curr = destMap[curr].left
        } else {
            curr = destMap[curr].right
        }
        numSteps += 1
    }

    return numSteps
}

func gcd(a int, b int) int {
    for b != 0 {
        temp := b
        b = a % b
        a = temp
    }

    return a
}

func lcm(a int, b int) int {
    return (a * b) / gcd(a, b)
}

func lcmOfSlice(slice []int) int {
    result := slice[0]

    for _, num := range(slice) {
        result = lcm(result, num)
    }

    return result
}

func main() {
    filename := "d8"
    file, _ := os.Open(filename)
    scanner := bufio.NewScanner(file)

    var buffer []string
    for scanner.Scan() {
        buffer = append(buffer, scanner.Text())
    }

    steps := buffer[0]
    buffer = buffer[2:]

    destMap, startPositions := parseLines2(buffer)
    var allNumSteps []int
    for _, startPos := range(startPositions) {
        numSteps := getNumSteps3(steps, destMap, startPos)
        allNumSteps = append(allNumSteps, numSteps)
    }

    fmt.Println(allNumSteps)
    numSteps := lcmOfSlice(allNumSteps)
    fmt.Println(numSteps)
}
